File: gremlin-dotnet/src/Gremlin.Net/Driver/Messages/RequestMessage.cs
#region License

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#endregion

using System;
using System.Collections.Generic;
using Gremlin.Net.Process.Traversal;

namespace Gremlin.Net.Driver.Messages
{
    /// <summary>
    ///     The model for a 4.0 request message sent to the server.
    /// </summary>
    public class RequestMessage
    {
        private RequestMessage(string gremlin, Dictionary<string, object> fields)
        {
            Gremlin = gremlin ?? throw new ArgumentNullException(nameof(gremlin));
            Fields = fields;
            if (!Fields.ContainsKey(Tokens.ArgsLanguage))
            {
                Fields[Tokens.ArgsLanguage] = "gremlin-lang";
            }
        }

        /// <summary>
        ///     Gets the Gremlin query string.
        /// </summary>
        public string Gremlin { get; }

        /// <summary>
        ///     Gets the fields map containing language, g, parameters, and other options.
        /// </summary>
        public Dictionary<string, object> Fields { get; }

        /// <summary>
        ///     Returns a copy of this message with <paramref name="key"/> set to
        ///     <paramref name="value"/>, without mutating this instance (and therefore without
        ///     mutating the caller-owned message). Used to fill connection-level defaults onto
        ///     the outgoing request only.
        /// </summary>
        /// <param name="key">The field key to set on the copy.</param>
        /// <param name="value">The field value to set on the copy.</param>
        /// <returns>A new <see cref="RequestMessage"/> carrying the added field.</returns>
        internal RequestMessage CloneWithField(string key, object value)
        {
            var copiedFields = new Dictionary<string, object>(Fields) { [key] = value };
            return new RequestMessage(Gremlin, copiedFields);
        }

        /// <summary>
        ///     Initializes a <see cref="Builder" /> to build a <see cref="RequestMessage" />.
        /// </summary>
        /// <param name="gremlin">The Gremlin query string.</param>
        /// <returns>A <see cref="Builder" /> to build a <see cref="RequestMessage" />.</returns>
        public static Builder Build(string gremlin)
        {
            return new Builder(gremlin);
        }

        /// <summary>
        ///     Allows to build <see cref="RequestMessage" /> objects.
        /// </summary>
        public class Builder
        {
            private readonly string _gremlin;
            private readonly Dictionary<string, object> _fields = new Dictionary<string, object>();
            private readonly Dictionary<string, object> _parameters = new Dictionary<string, object>();
            private string? _parametersString;

            internal Builder(string gremlin)
            {
                _gremlin = gremlin;
            }

            /// <summary>
            ///     Sets the traversal source name.
            /// </summary>
            /// <param name="g">The traversal source name.</param>
            /// <returns>The <see cref="Builder" />.</returns>
            public Builder AddG(string g)
            {
                _fields[Tokens.ArgsG] = g;
                return this;
            }

            /// <summary>
            ///     Adds a single query parameter.
            /// </summary>
            /// <param name="key">The parameter key.</param>
            /// <param name="val">The parameter value.</param>
            /// <returns>The <see cref="Builder" />.</returns>
            public Builder AddParameter(string key, object val)
            {
                if (_parametersString != null)
                    throw new InvalidOperationException("Cannot mix AddParameter() with AddParametersString().");
                _parameters[key] = val;
                return this;
            }

            /// <summary>
            ///     Adds multiple query parameters from a dictionary. The values will be
            ///     converted to a gremlin-lang string when the message is created.
            ///     Cannot be mixed with <see cref="AddParametersString"/>.
            /// </summary>
            /// <param name="parameters">The parameters to add.</param>
            /// <returns>The <see cref="Builder" />.</returns>
            public Builder AddParameters(Dictionary<string, object> parameters)
            {
                if (_parametersString != null)
                    throw new InvalidOperationException("Cannot mix AddParameters() with AddParametersString().");
                foreach (var kvp in parameters)
                {
                    _parameters[kvp.Key] = kvp.Value;
                }
                return this;
            }

            /// <summary>
            ///     Sets the query parameters as a pre-serialized gremlin-lang map literal string.
            ///     Cannot be mixed with <see cref="AddParameter"/> or <see cref="AddParameters"/>.
            /// </summary>
            /// <param name="parametersString">The gremlin-lang parameters string.</param>
            /// <returns>The <see cref="Builder" />.</returns>
            public Builder AddParametersString(string parametersString)
            {
                if (_parameters.Count > 0)
                    throw new InvalidOperationException("Cannot mix AddParametersString() with AddParameter()/AddParameters().");
                _parametersString = parametersString;
                return this;
            }

            /// <summary>
            ///     Adds a field to the request message.
            /// </summary>
            /// <param name="key">The field key.</param>
            /// <param name="value">The field value.</param>
            /// <returns>The <see cref="Builder" />.</returns>
            public Builder AddField(string key, object value)
            {
                _fields[key] = value;
                return this;
            }

            /// <summary>
            ///     Checks whether a field has been set.
            /// </summary>
            /// <param name="key">The field key to check.</param>
            /// <returns>True if the field exists.</returns>
            public bool HasField(string key) => _fields.ContainsKey(key);

            /// <summary>
            ///     Sets the evaluation timeout for this request.
            /// </summary>
            /// <param name="timeout">The timeout value.</param>
            /// <returns>The <see cref="Builder" />.</returns>
            public Builder AddEvaluationTimeout(object timeout)
            {
                _fields[Tokens.ArgsEvalTimeout] = timeout;
                return this;
            }

            /// <summary>
            ///     Sets the batch size for this request.
            /// </summary>
            /// <param name="batchSize">The batch size value.</param>
            /// <returns>The <see cref="Builder" />.</returns>
            public Builder AddBatchSize(object batchSize)
            {
                _fields[Tokens.ArgsBatchSize] = batchSize;
                return this;
            }

            /// <summary>
            ///     Sets the materializeProperties option for this request.
            /// </summary>
            /// <param name="materializeProperties">The materializeProperties value (e.g. "all" or "tokens").</param>
            /// <returns>The <see cref="Builder" />.</returns>
            public Builder AddMaterializeProperties(string materializeProperties)
            {
                _fields[Tokens.ArgMaterializeProperties] = materializeProperties;
                return this;
            }

            /// <summary>
            ///     Sets the user agent for this request.
            /// </summary>
            /// <param name="userAgent">The user agent string.</param>
            /// <returns>The <see cref="Builder" />.</returns>
            public Builder AddUserAgent(string userAgent)
            {
                _fields[Tokens.ArgsUserAgent] = userAgent;
                return this;
            }

            /// <summary>
            ///     Sets the bulk results option for this request.
            /// </summary>
            /// <param name="bulkResults">The bulk results value.</param>
            /// <returns>The <see cref="Builder" />.</returns>
            public Builder AddBulkResults(object bulkResults)
            {
                _fields[Tokens.ArgsBulkResults] = bulkResults;
                return this;
            }

            /// <summary>
            ///     Sets the language for this request.
            /// </summary>
            /// <param name="language">The language identifier (e.g. "gremlin-lang").</param>
            /// <returns>The <see cref="Builder" />.</returns>
            public Builder AddLanguage(string language)
            {
                _fields[Tokens.ArgsLanguage] = language;
                return this;
            }

            /// <summary>
            ///     Creates the <see cref="RequestMessage" /> given the settings provided to the <see cref="Builder" />.
            /// </summary>
            /// <returns>The built <see cref="RequestMessage" />.</returns>
            public RequestMessage Create()
            {
                // prefer pre-serialized parameters string over raw map
                if (_parametersString != null)
                {
                    _fields[Tokens.ArgsParameters] = _parametersString;
                }
                else if (_parameters.Count > 0)
                {
                    _fields[Tokens.ArgsParameters] = GremlinLang.ConvertParametersToString(_parameters);
                }
                return new RequestMessage(_gremlin, _fields);
            }
        }
    }
}

File: gremlin-util/src/main/java/org/apache/tinkerpop/gremlin/driver/message/RequestMessage.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.tinkerpop.gremlin.driver.message;

import org.apache.tinkerpop.gremlin.driver.Tokens;
import org.apache.tinkerpop.gremlin.structure.util.ElementHelper;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

/**
 * The model for a request message sent to the server.
 *
 * @author Stephen Mallette (http://stephen.genoprime.com)
 */
public final class RequestMessage {
    /**
     * An "invalid" message.  Used internally only.
     */
    public static final RequestMessage INVALID = new RequestMessage(Tokens.OPS_INVALID);

    private final UUID requestId;
    private final String op;
    private final String processor;
    private final Map<String, Object> args;

    private RequestMessage(final UUID requestId, final String op, final String processor, final Map<String, Object> args) {
        this.requestId = requestId;
        this.op = op;
        this.processor = processor;
        this.args = Optional.ofNullable(args).orElse(new HashMap<>());
    }

    /**
     * Empty constructor for serialization.
     */
    private RequestMessage() {
        this(null);
    }

    private RequestMessage(final String op) {
        this(null, op, null, null);
    }

    /**
     * The id of the current request and is used to track the message within Gremlin Server and in its response.  This
     * value should be unique per request made.
     */
    public UUID getRequestId() {
        return requestId;
    }

    /**
     * The operation or command to perform as defined by a particular Processor.
     */
    public String getOp() {
        return op;
    }

    /**
     * The name of the Processor that should handle the {@link #op}.  Defaults to the standard processor if
     * not specified.
     */
    public String getProcessor() {
        return processor;
    }

    /**
     * A {@link Map} of arguments that are supplied to the {@link #op}.  Each {@link #op} accepts different argument,
     * so consult the documentation for a particular one to understand what is expected.
     */
    public Map<String, Object> getArgs() {
        return args;
    }

    public <T> Optional<T> optionalArgs(final String key) {
        final Object o = args.get(key);
        return o == null ? Optional.empty() : Optional.of((T) o);
    }

    public static Builder from(final RequestMessage msg) {
        final Builder builder = build(msg.op)
                .overrideRequestId(msg.requestId)
                .processor(msg.processor);
        msg.args.forEach(builder::addArg);
        return builder;
    }

    public static Builder build(final String op) {
        return new Builder(op);
    }

    /**
     * Builder class for {@link RequestMessage}.
     */
    public static final class Builder {
        public static final String OP_PROCESSOR_NAME = "";
        private UUID requestId;
        private String op;
        private String processor = OP_PROCESSOR_NAME;
        private Map<String, Object> args = new HashMap<>();

        private Builder(final String op) {
            this.op = op;
        }

        /**
         * If this value is not set in the builder then the {@link RequestMessage#processor} defaults to
         * the standard op processor (empty string).
         *
         * @param processor the name of the processor
         */
        public Builder processor(final String processor) {
            this.processor = processor;
            return this;
        }

        /**
         * Override the request identifier with a specified one, otherwise the {@link Builder} will randomly generate
         * a {@link UUID}.
         */
        public Builder overrideRequestId(final UUID requestId) {
            this.requestId = requestId;
            return this;
        }

        public Builder addArg(final String key, final Object val) {
            args.put(key, val);
            return this;
        }

        public Builder add(final Object... keyValues) {
            args.putAll(ElementHelper.asMap(keyValues));
            return this;
        }

        /**
         * Create the request message given the settings provided to the {@link Builder}.
         */
        public RequestMessage create() {
            return new RequestMessage(null == requestId ? UUID.randomUUID() : requestId, op, processor, args);
        }
    }

    @Override
    public String toString() {
        return "RequestMessage{" +
                ", requestId=" + requestId +
                ", op='" + op + '\'' +
                ", processor='" + processor + '\'' +
                ", args=" + args +
                '}';
    }
}
